BUG: Slice Arrow buffer before passing it to numpy (#40896)

ThomasBlauthQC opened the pull request:
Add Arrow buffer slicing before handing it over to numpy which is
needed in case the Arrow buffer contains padding or offset.

- [x] closes #40896
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Points I'm not sure about:
- I created a new test file. Would it be better to place the test somewhere else?
- The test only tests int32. Should I add more dtypes?

I'm still new to the internals of pandas and always trying to improve so feedback of any kind is highly appreciated!

ThomasBlauthQC committed:
BUG: Slice Arrow buffer before passing it to numpy (#40896)

Add Arrow buffer slicing before handing it over to numpy which is
needed in case the Arrow buffer contains padding or offset.

ThomasBlauthQC committed:
BUG: Slice Arrow buffer before passing it to numpy (#40896)

Use numpy dtype for bitwidth information.
Fix test signature.

ThomasBlauthQC committed:
BUG: Slice Arrow buffer before passing it to numpy (#40896)

Fix mask buffer in test.

jorisvandenbossche commented:
Thanks for the PR!

> I created a new test file. Would it be better to place the test somewhere else?

There is a `pandas/tests/arrays/masked/test_arrow_compat.py`. I think you can put it there.

> The test only tests int32. Should I add more types?

Yes, I think it would be good to parameterize the test across different dtypes.

jorisvandenbossche commented on pandas/core/arrays/_arrow_utils.py:
```diff
@@ -25,8 +27,16 @@ def pyarrow_array_to_numpy_and_mask(arr, dtype):
         Tuple of two numpy arrays with the raw data (with specified dtype) and
         a boolean mask (validity mask, so False means missing)
     """
+    dtype = np.dtype(dtype)
```
This should not be needed, as we can ensure internally to always pass a numpy dtype object. I see there is one occurrence in `PeriodDtype` where we pass a string type -> you can update that case to pass `np.dtype("int64")` instead.

ThomasBlauthQC replied:
Yes, that was exactly the reason I added the line. I changed `PeriodDtype` the way you suggested.

Unfortunately some other tests break when I remove the line. IIUC pyarrow is using the function as well.

```shell
______________________________________________ test_arrow_sliced _____________________________________________________________________________________

    @td.skip_if_no("pyarrow", min_version="0.16.0")
    def test_arrow_sliced():
        # https://github.com/pandas-dev/pandas/issues/38525
        import pyarrow as pa
    
        df = pd.DataFrame({"a": pd.array([0, None, 2, 3, None], dtype="Int64")})
        table = pa.table(df)
>       result = table.slice(2, None).to_pandas()

pandas/tests/arrays/masked/test_arrow_compat.py:65: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pyarrow/array.pxi:751: in pyarrow.lib._PandasConvertible.to_pandas
    ???
pyarrow/table.pxi:1668: in pyarrow.lib.Table._to_pandas
    ???
/opt/conda/lib/python3.8/site-packages/pyarrow/pandas_compat.py:792: in table_to_blockmanager
    blocks = _table_to_blocks(options, table, categories, ext_columns_dtypes)
/opt/conda/lib/python3.8/site-packages/pyarrow/pandas_compat.py:1133: in _table_to_blocks
    return [_reconstruct_block(item, columns, extension_columns)
/opt/conda/lib/python3.8/site-packages/pyarrow/pandas_compat.py:1133: in <listcomp>
    return [_reconstruct_block(item, columns, extension_columns)
/opt/conda/lib/python3.8/site-packages/pyarrow/pandas_compat.py:751: in _reconstruct_block
    pd_ext_arr = pandas_dtype.__from_arrow__(arr)
pandas/core/arrays/numeric.py:65: in __from_arrow__
    data, mask = pyarrow_array_to_numpy_and_mask(arr, dtype=self.type)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

arr = <pyarrow.lib.Int64Array object at 0x7f77e8678820>
[
  2,
  3,
  null
], dtype = <class 'numpy.int64'>

    def pyarrow_array_to_numpy_and_mask(arr, dtype):
        """
        Convert a primitive pyarrow.Array to a numpy array and boolean mask based
        on the buffers of the Array.
    
        At the moment pyarrow.BooleanArray is not supported.
    
        Parameters
        ----------
        arr : pyarrow.Array
        dtype : numpy.dtype
    
        Returns
        -------
        (data, mask)
            Tuple of two numpy arrays with the raw data (with specified dtype) and
            a boolean mask (validity mask, so False means missing)
        """
        # dtype = np.dtype(dtype)
        buflist = arr.buffers()
        # Since Arrow buffers might contain padding and the data might be offset,
        # the buffer gets sliced here before handing it to numpy.
        # See also https://github.com/pandas-dev/pandas/issues/40896
>       offset = arr.offset * dtype.itemsize
E       TypeError: unsupported operand type(s) for *: 'int' and 'getset_descriptor'

pandas/core/arrays/_arrow_utils.py:35: TypeError
```

ThomasBlauthQC replied:
> This should not be needed, as we can ensure internally to always pass a numpy dtype object. I see there is one occurrence in `PeriodDtype` where we pass a string type -> you can update that case to pass `np.dtype("int64")` instead.

The change form `"int64"` to `np.dtype("int64")` caused some trouble with timezone conversion on Windows. Revertig it lets the tests pass.
```shell
Traceback (most recent call last):
  File "pandas\_libs\tslibs\tzconversion.pyx", line 385, in pandas._libs.tslibs.tzconversion.tz_convert_utc_to_tzlocal
    return _tz_convert_tzlocal_utc(utc_val, tz, to_utc=False, fold=fold)
  File "pandas\_libs\tslibs\tzconversion.pyx", line 596, in pandas._libs.tslibs.tzconversion._tz_convert_tzlocal_utc
    delta = _tzlocal_get_offset_components(val, tz, to_utc, fold)
  File "pandas\_libs\tslibs\tzconversion.pyx", line 556, in pandas._libs.tslibs.tzconversion._tzlocal_get_offset_components
    dt = dt.astimezone(tz)
  File "C:\Miniconda\envs\pandas-dev\lib\site-packages\dateutil\tz\_common.py", line 144, in fromutc
    return f(self, dt)
  File "C:\Miniconda\envs\pandas-dev\lib\site-packages\dateutil\tz\_common.py", line 261, in fromutc
    _fold = self._fold_status(dt, dt_wall)
  File "C:\Miniconda\envs\pandas-dev\lib\site-packages\dateutil\tz\_common.py", line 196, in _fold_status
    if self.is_ambiguous(dt_wall):
  File "C:\Miniconda\envs\pandas-dev\lib\site-packages\dateutil\tz\tz.py", line 254, in is_ambiguous
    naive_dst = self._naive_is_dst(dt)
  File "C:\Miniconda\envs\pandas-dev\lib\site-packages\dateutil\tz\tz.py", line 260, in _naive_is_dst
    return time.localtime(timestamp + time.timezone).tm_isdst
OSError: [Errno 22] Invalid argument
```

jorisvandenbossche replied:
OK, that's a good reason to keep it then!

ThomasBlauthQC committed:
BUG: Slice Arrow buffer before passing it to numpy (#40896)

Move tests to pandas/tests/arrays/masked/test_arrow_compat.py
Add more dtypes to test.

ThomasBlauthQC commented:
I moved the test and added parameterization for all numeric types.

jreback commented on pandas/tests/arrays/masked/test_arrow_compat.py:
```diff
@@ -64,3 +65,102 @@ def test_arrow_sliced():
+
+@pytest.fixture
+def np_dtype_to_arrays(request):
+    import pyarrow as pa
```
can we just add an import_optional_dependency at the top of this file, rather than having skips on every test.

ThomasBlauthQC replied:
I used pyarrow 0.16.0 as the minimal version so `test_arrow_array` no longer gets run with pyarrow 0.15.0. Hope this is appropriate.

jreback commented on pandas/tests/arrays/masked/test_arrow_compat.py:
```diff
@@ -64,3 +65,102 @@ def test_arrow_sliced():
+
+@td.skip_if_no("pyarrow")
+@pytest.mark.parametrize(
+    "np_dtype_to_arrays",
```
can you use the fixture: any_real_dtype or any_numpy_dtype instead

ThomasBlauthQC committed:
BUG: Slice Arrow buffer before passing it to numpy (#40896)

Modify test_arrow_compat.py:
- Use import_optional_dependency to skip the whole module if
  pyarrow is not available.
- Use any_real_dtype fixture.

jorisvandenbossche commented on pandas/tests/arrays/masked/test_arrow_compat.py:
```diff
@@ -15,10 +21,8 @@ def data(request):
     return request.param
 
 
-@td.skip_if_no("pyarrow", min_version="0.15.0")
```
Note that this one is actually already tested for 0.15.0 as well, while the others only for 0.16. As long as 0.15 is the minimal supported pyarrow version, I think we should keep the individual skips.

jorisvandenbossche commented on pandas/tests/arrays/masked/test_arrow_compat.py:
```diff
@@ -1,10 +1,16 @@
+try:
+    pa = import_optional_dependency("pyarrow", min_version="0.16.0")
+except ImportError:
+    pytestmark = pytest.mark.skip(reason="Pyarrow not available")
```
To avoid the imports in every test (as you did now), you could also do a 

```
try:
    import pyarrow as pa
except ImportError:
    pa = None
```

jorisvandenbossche commented on pandas/tests/arrays/masked/test_arrow_compat.py:
```diff
@@ -54,13 +54,87 @@ def test_arrow_from_arrow_uint():
+    np_dtype = np.dtype(any_real_dtype)
+    pa_type = pa.from_numpy_dtype(np_dtype)
+
+    pa_array = pa.array([0, 1, 2], type=pa_type)
```
```suggestion
    pa_array = pa.array([0, 1, 2, None], type=pa_type)
```

Maybe good to include a missing value, to ensure there is a bitmask buffer?

ThomasBlauthQC replied:
Yes, good point!

ThomasBlauthQC committed:
BUG: Slice Arrow buffer before passing it to numpy (#40896)

Add whatsnew entry.

ThomasBlauthQC committed:
BUG: Slice Arrow buffer before passing it to numpy (#40896)

Enforce creation of Arrow bitmask buffer.

ThomasBlauthQC committed:
BUG: Slice Arrow buffer before passing it to numpy (#40896)

Merge branch 'master' into issue-40896

ThomasBlauthQC committed:
BUG: Slice Arrow buffer before passing it to numpy (#40896)

Fix whatsnew.rst

jreback commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -882,6 +882,7 @@ ExtensionArray
 - Fixed bug where :meth:`Series.idxmax`, :meth:`Series.idxmin` and ``argmax/min`` fail when the underlying data is :class:`ExtensionArray` (:issue:`32749`, :issue:`33719`, :issue:`36566`)
 - Fixed a bug where some properties of subclasses of :class:`PandasExtensionDtype` where improperly cached (:issue:`40329`)
 - Bug in :meth:`DataFrame.mask` where masking a :class:`Dataframe` with an :class:`ExtensionArray` dtype raises ``ValueError`` (:issue:`40941`)
+- Bug in :func:`pyarrow_array_to_numpy_and_mask` where Numpy raises ``ValueError`` if buffer size does not match multiple of dtype size (:issue:`40896`)
```
can you change this to a user facing note.

jreback commented on pandas/tests/arrays/masked/test_arrow_compat.py:
```diff
@@ -1,10 +1,16 @@
 import pandas as pd
 import pandas._testing as tm
 
+try:
```
just use pyimportorskip here. i am not sure we can test anything w/o it.

ThomasBlauthQC committed:
Fix whatsnew

ThomasBlauthQC committed:
Use pytest.importorskip

jorisvandenbossche commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -882,6 +882,7 @@ ExtensionArray
 - Fixed bug where :meth:`Series.idxmax`, :meth:`Series.idxmin` and ``argmax/min`` fail when the underlying data is :class:`ExtensionArray` (:issue:`32749`, :issue:`33719`, :issue:`36566`)
 - Fixed a bug where some properties of subclasses of :class:`PandasExtensionDtype` where improperly cached (:issue:`40329`)
 - Bug in :meth:`DataFrame.mask` where masking a :class:`Dataframe` with an :class:`ExtensionArray` dtype raises ``ValueError`` (:issue:`40941`)
+- Bug in :func:`pyarrow_array_to_numpy_and_mask` when converting a pyarrow array who's data buffer size is not a multiple of dtype size (:issue:`40896`)
```
```suggestion
- Bug in the conversion from pyarrow to pandas (e.g. for reading Parquet) with nullable dtypes and a pyarrow array whose data buffer size is not a multiple of dtype size (:issue:`40896`)
```

`pyarrow_array_to_numpy_and_mask` is not a public method, so we should try to avoid mentioning it. So a suggested rewording above. 

Also, I would maybe move this to the `I/O` section

ThomasBlauthQC replied:
I changed that as suggested. Also, let me know if I should rebase to get rid of the small commits.

jorisvandenbossche replied:
> Also, let me know if I should rebase to get rid of the small commits.

No need to do that, we will squash everything when merging (well, github does that for us ;))

ThomasBlauthQC committed:
Fix whatsnew

jorisvandenbossche reviewed (approved):
Thanks for the updates! Looking good to me

jreback commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -796,6 +796,7 @@ I/O
 - Bug in :func:`read_excel` dropping empty values from single-column spreadsheets (:issue:`39808`)
 - Bug in :meth:`DataFrame.to_string` misplacing the truncation column when ``index=False`` (:issue:`40907`)
 - Bug in :func:`read_orc` always raising ``AttributeError`` (:issue:`40918`)
+- Bug in the conversion from pyarrow to pandas (e.g. for reading Parquet) with nullable dtypes and a pyarrow array whose data buffer size is not a multiple of dtype size (:issue:`40896`)
```
would not object to actually showing `read_parquet` reference here

jorisvandenbossche replied:
I don't think that's needed (the original bug report didn't actually involve a plain reading of a Parquet file (not sure this can happen with that), but with a dataset scan+filter operation (from any source format).

ThomasBlauthQC replied:
Yes, I was creating the arrow table using only the arrow api. The bug then occured on the conversion to pandas.

I did not actually try to reproduce the bug using `read_parquet`, so I'm also not sure if it can happen this way.

jorisvandenbossche replied:
I suppose it could happen when reading a parquet file using the new dataset API by passing a certain filter. But anyway, I don't think we need to make that explicit, IMO the above error message is clear enough.

jreback commented on pandas/tests/arrays/masked/test_arrow_compat.py:
```diff
@@ -89,12 +86,89 @@ def test_arrow_sliced(data):
 
 
+@pytest.fixture
+def np_dtype_to_arrays(any_real_dtype):
```
can you type this (mainly the output for better readability) also adding a doc-string explaining the purpose

jorisvandenbossche replied:
This is a fixture in the tests, I don't think we generally ask for annotations there? (eg in conftest.py there is no typing)

jreback commented on pandas/tests/arrays/masked/test_arrow_compat.py:
```diff
@@ -89,12 +86,89 @@ def test_arrow_sliced(data):
+
+
+def test_pyarrow_array_to_numpy_and_mask(np_dtype_to_arrays):
+    """
```
add the issue reference as a comment

jorisvandenbossche replied:
The issue reference is already included a few lines below

jreback commented on pandas/tests/arrays/masked/test_arrow_compat.py:
```diff
@@ -89,12 +86,89 @@ def test_arrow_sliced(data):
+    Also tests empty pyarrow arrays with non empty buffers.
+    See https://github.com/pandas-dev/pandas/issues/40896
+    """
+    from pandas.core.arrays._arrow_utils import pyarrow_array_to_numpy_and_mask
```
can you import at the top (after the pyarrow check is fine)

ThomasBlauthQC replied:
Should I move the rest of the pandas imports below the pyarrow check as well? Looks a bit weird having them seperated by the check.

ThomasBlauthQC committed:
Move import to top

jreback commented:
thanks @ThomasBlauthQC 

i think the checks are running, but can merge on green.

mlondschien commented:
Looks green @jreback. Merge?

jorisvandenbossche commented:
@ThomasBlauthQC Thanks a lot for the fix!
